source/common/http/http1/codec_impl.cc

```diff
@@ -174,8 +174,10 @@ void StreamEncoderImpl::encodeHeadersBase(const RequestOrResponseHeaderMap& head
   if (saw_content_length || disable_chunk_encoding_) {
     chunk_encoding_ = false;
   } else {
-    if (status && *status == 100) {
+    if (status && (*status == 100 || *status == 304)) {
```

Comment:
I put 304 here because it also needs to account for head request which the 204 code path seems to ignore.

Reply:
Thanks for picking this up. We've got a function for requests, requestShouldHaveNoBody, which handles this corner case on the request path.  I wonder if we could have a similar function (and input into bodiless requests) on the response path?  If so I think we'd also need to ensure there was no body when receiving incoming 304s, and runtime guard per the contribution guidelines.

We'll definitely want to runtime guard, per

Reply:
@alyssawilk I did think about making an option to change this behavior but found out that envoy also add Content-Length: 0 to the 304 response if origin doesn't provide one which I think is incorrect behavior.  I can look into adding an option to switch this behavior if needed.

Reply:
ok. Looking at envoy.reloadable_features.strict_1xx_and_204_response_headers, I understand better now and will do similar runtime guard. Should I also make a responseShouldHaveNoBody() that return true for 1xx, 204 and 304 status code response instead of checking for status code for 304 explicitly?

Reply:
Um... while 1xx, 204 and 304 all should not have body but 304 is the odd ball here where non-zero Content-Length is actually allowed by the spec. So, a single responseShouldHaveNoBody() function is not enough for 304.

Reply:
Ah, that's a good call, sorry for not thinking that one out.  If we need to just special case this here, I'm fine with it. :-)

We will need to runtime guard per CONTRIBUTING.md, given it's a user-visible data plane change, even if it's a bugfix, and then if we can flip that (with TestScopedRuntime) in one codec unit test that'd be great.

Reply:
yes. I will definitely do the runtime guard. Also adding more tests in protocol_integration_test as suggested by @antoniovicente . Confirmed HTTP/2 behaves as I would expected without any changes. For HTTP/1, here is the odd result before the change:
1) GET request and upstream returns 304 without Content-Length. "Content-Length: 0" is added to the downstream response
2) HEAD request and upstream returns 304 without Content-Length. Both "Content-Length: 0" and "Transfer-Encoding: chunked" are added to the downstream response.